src/Evolution/Systems/ScalarWave/BoundaryCorrections/UpwindPenalty.cpp

```diff
@@ -0,0 +1,224 @@
+    dg::Formulation /*dg_formulation*/) const noexcept {
+  const size_t num_pts = char_speeds_int[0].size();
+  Variables<tmpl::list<::Tags::TempScalar<0>, ::Tags::TempScalar<1>,
+                       ::Tags::TempScalar<2>, ::Tags::TempScalar<3>,
+                       ::Tags::TempScalar<4>, ::Tags::TempScalar<5>,
+                       ::Tags::TempScalar<6>, ::Tags::TempScalar<7>>>
+      buffer(num_pts);
+  DataVector& weighted_lambda_psi_int = get(get<::Tags::TempScalar<0>>(buffer));
+  weighted_lambda_psi_int = step_function(-char_speeds_int[0]);
+  DataVector& weighted_lambda_psi_ext = get(get<::Tags::TempScalar<1>>(buffer));
+  weighted_lambda_psi_ext = -step_function(char_speeds_ext[0]);
+
+  DataVector& weighted_lambda_zero_int =
+      get(get<::Tags::TempScalar<2>>(buffer));
+  weighted_lambda_zero_int = step_function(-char_speeds_int[0]);
+  DataVector& weighted_lambda_zero_ext =
+      get(get<::Tags::TempScalar<3>>(buffer));
+  weighted_lambda_zero_ext = -step_function(char_speeds_ext[0]);
+
+  DataVector& weighted_lambda_plus_int =
+      get(get<::Tags::TempScalar<4>>(buffer));
+  weighted_lambda_plus_int = step_function(-char_speeds_int[1]);
+  DataVector& weighted_lambda_plus_ext =
+      get(get<::Tags::TempScalar<5>>(buffer));
+  weighted_lambda_plus_ext = -step_function(char_speeds_ext[1]);
+
+  DataVector& weighted_lambda_minus_int =
+      get(get<::Tags::TempScalar<6>>(buffer));
+  weighted_lambda_minus_int = step_function(-char_speeds_int[2]);
+  DataVector& weighted_lambda_minus_ext =
+      get(get<::Tags::TempScalar<7>>(buffer));
+  weighted_lambda_minus_ext = -step_function(char_speeds_ext[2]);
+
+  // D_psi = Theta(-lambda_psi^{ext}) lambda_psi^{ext} v_psi^{ext}
+  //       - Theta(-lambda_psi^{int}) lambda_psi^{int} v_psi^{int}
+  // where the unit normals on both sides point in the same direction, out
+  // of the current element. Since lambda_psi from the neighbor is computing
+  // with the normal vector pointing into the current element in the code,
+  // we need to swap the sign of lambda_psi^{ext}. Theta is the heaviside step
+  // function with Theta(0) = 0.
+  psi_boundary_correction->get() =
+      weighted_lambda_psi_ext * get(char_speed_v_psi_ext) -
+      weighted_lambda_psi_int * get(char_speed_v_psi_int);
+
+  get(*pi_boundary_correction) =
+      0.5 * (weighted_lambda_plus_ext * get(char_speed_v_plus_ext) +
+             weighted_lambda_minus_ext * get(char_speed_v_minus_ext)) +
+      weighted_lambda_psi_ext * get(char_speed_constraint_gamma2_v_psi_ext)
+
+      - 0.5 * (weighted_lambda_plus_int * get(char_speed_v_plus_int) +
+               weighted_lambda_minus_int * get(char_speed_v_minus_int)) -
+      weighted_lambda_psi_int * get(char_speed_constraint_gamma2_v_psi_int);
+
+  for (size_t d = 0; d < Dim; ++d) {
+    // Overall minus sign on ext because of normal vector is opposite direction.
+    phi_boundary_correction->get(d) =
+        -0.5 * (weighted_lambda_minus_ext *
```

Comment:
should this extra sign also appear in the final formula in the documentation where it notes the sign flips?

Reply:
Yep!

Reply:
Wait, this is in agreement with the documentation as far as I can tell.